src/ploneintranet/workspace/adapters.py

```diff
@@ -88,3 +98,270 @@ def getRoles(self, principal_id):
+                'john-doe': set([uid, uid]),
+                'jane-doe': set([uid]),
+            }
+            'type': {
+                'foo': set([uid]),
+                'bar': set([uid, uid, uid]),
+                'baz': set([uid]),
+            }
+            'first_letter': {
+                'a': set([uid]),
+                'b': set([uid, uid, uid]),
+                'c': set([uid]),
+            }
+        })
+        The top-level keys are the groupings.
+
+        For each grouping we store another dict.
+
+        Each key in this dict is a unique value for that grouping. These values
+        are retrieved from the objects stored in the workspace.
+
+        For each key we have a list of uids. These are the uids of the objects
+        that have that key as a field value (corresponding to the grouping).
+
+        Remember: each grouping is a field on an object, so each value of that
+        grouping is a value of that field on an object inside the workspace.
+
+        We need to keep track of the uids, so that we know when to remove a
+        grouping-value. When an object is modified, we don't know if anything
+        was removed (for example from the 'Subject' field, which corresponds to
+        'label' grouping).
+
+        So we have to check every time if that object's uid is in any grouping
+        values that that object doesn't have anymore. In that way, we know to
+        remove the uid from that grouping-value. If that grouping-value doesn't
+        have any uids anymore, we can remove it.
+    """
+
+    def __init__(self, context):
+        self.context = context
+        context = aq_inner(self.context)
+        if not hasattr(context, '_groupings'):
+            self.init_groupings()
+
+    def clear_groupings(self):
+        self.init_groupings()
+
+    def init_groupings(self):
+        context = aq_inner(self.context)
+        context._groupings = OOBTree({
+            'label': OrderedBTreeFolderBase(),
+            'author': OrderedBTreeFolderBase(),
+            'type': OrderedBTreeFolderBase(),
+            'first_letter': OrderedBTreeFolderBase(),
+        })
+
+    def _add_grouping_values(self, grouping, values, obj):
+        """ Add $uid to the list of uids stored under the grouping values
+            (provided by $values) in the groupings datastructure.
+
+            If the list doesn't exist yet, add it.
+        """
+        uid = IUUID(obj)
+        groupings = self.get_groupings()
+        for value in values:
+            if value in groupings[grouping]:
+                if uid in groupings[grouping][value]:
+                    continue
+                else:
+                    groupings[grouping][value].add(uid)
+                    groupings[grouping][value]._p_changed = 1
+            else:
+                groupings[grouping][value] = GroupingStorageValues([uid])
+
+        groupings[grouping]._p_changed = 1
+        groupings._p_changed = 1
```

Comment:
I don't like the sight of these ._p_changed. It's brittle code and if you somehow miss one of the levels you're in big trouble. If you're using a Persistent storage object and you're mutating that object, you shouldn't then have to tell that object that it was mutated. That should be an internal of the storage object, not something you need to manage from the calling code.

In my own code I've used OOBTree.OOTreeSet and the like for similar storage purposes. 
https://pythonhosted.org/BTrees/

That gives you two benefits: 
1) no need to manually manage ._p_changed 
2) use the .insert() method to do away with the nested if/if/else/else state checks. Insert inserts a value if it is not there, but *does nothing* if the value is already there, meaning you can safely use it to *always* initialize a datastructure without risking existing data getting lost.

See e.g.
https://github.com/ploneintranet/ploneintranet/blob/master/src/ploneintranet/network/graph.py#L27